src/SqlSquared/Parser.purs

```diff
@@ -382,74 +323,62 @@ cases = PC.try do
 
-unshiftExpr
-  ∷ ∀ t m
-  . Corecursive t (Sig.SqlF EJ.EJsonF)
-  ⇒ Monad m
-  ⇒ P.ParserT (Array Token) m t
-unshiftExpr =
-  (PC.try do
-      _ ← operator "["
-      e ← expr
-      _ ← operator "..."
-      _ ← operator "]"
-      pure $ embed $ Sig.Unop { op: Sig.UnshiftArray, expr: e })
-  <|> (PC.try do
-          _ ← operator "{"
-          lhs ← expr
-          _ ← operator ":"
-          rhs ← expr
-          _ ← operator "..."
-          _ ← operator "}"
-          pure $ embed $ Sig.Binop { op: Sig.UnshiftMap, lhs, rhs })
-
-parenList
-  ∷ ∀ t m
-  . Corecursive t (Sig.SqlF EJ.EJsonF)
-  ⇒ Monad m
-  ⇒ P.ParserT (Array Token) m (L.List t)
-parenList = PC.try do
+unshiftExpr ∷ ∀ m t. SqlParser' m t
+unshiftExpr = unshiftArrayExpr <|> unshiftMapExpr
+  where
+  unshiftArrayExpr = do
+    _ ← operator "["
+    e ← expr
+    _ ← operator "..."
+    _ ← operator "]"
+    pure $ embed $ Sig.Unop { op: Sig.UnshiftArray, expr: e }
+
+  unshiftMapExpr = do
+    _ ← operator "{"
+    lhs ← expr
+    _ ← operator ":"
+    rhs ← expr
+    _ ← operator "..."
+    _ ← operator "}"
+    pure $ embed $ Sig.Binop { op: Sig.UnshiftMap, lhs, rhs }
+
+parenList ∷ ∀ m t. SqlParser m t (L.List t)
+parenList = do
   operator "("
   arr ← PC.sepBy expr $ operator ","
   operator ")"
   pure $ L.fromFoldable arr
 
-unaryOperator
-  ∷ ∀ t m
-  . Corecursive t (Sig.SqlF EJ.EJsonF)
-  ⇒ Monad m
-  ⇒ P.ParserT (Array Token) m t
-unaryOperator = PC.try do
-  op ←
-    (operator "-" $> Sig.Negative)
-    <|> (operator "+" $> Sig.Positive)
-    <|> (keyword "distinct" $> Sig.Distinct)
-    <|> (keyword "not" $> Sig.Not)
-    <|> (keyword "exists" $> Sig.Exists)
-  e ← primaryExpression
-  pure $ embed $ Sig.Unop { op, expr: e}
-
-functionExpr
-  ∷ ∀ t m
-  . Corecursive t (Sig.SqlF EJ.EJsonF)
-  ⇒ Monad m
-  ⇒ P.ParserT (Array Token) m t
+unaryOperator ∷ ∀ m t. SqlParser' m t
+unaryOperator = do
+  op ← unaryOp
+  e ← primaryExpr
+  pure $ embed $ Sig.Unop { op, expr: e }
+  where
+  unaryOp = PC.choice
+    [ operator "-" $> Sig.Negative
+    , operator "+" $> Sig.Positive
+    , keyword "distinct" $> Sig.Distinct
+    , keyword "not" $> Sig.Not
+    , keyword "exists" $> Sig.Exists
+    ]
+
+functionExpr ∷ ∀ m t. SqlParser' m t
 functionExpr = PC.try do
   name ← ident
   args ← parenList
-  pure $ embed $ Sig.InvokeFunction {name, args}
+  pure $ embed $ Sig.InvokeFunction { name, args }
 
 functionDecl
   ∷ ∀ m a
   . Monad m
-  ⇒ P.ParserT (Array Token) m a
-  → P.ParserT (Array Token) m (Sig.SqlDeclF a)
-functionDecl parseExpr = PC.try do
-  _ ← keyword "create"
-  _ ← keyword "function"
+  ⇒ P.ParserT TokenStream m a
+  → P.ParserT TokenStream m (Sig.SqlDeclF a)
+functionDecl parseExpr = asErrorMessage "function declaration" do
+  _ ← PC.try $ keyword "create" *> keyword "function"
   name ← ident
   operator "("
-  args ← PC.sepBy (operator ":" *> ident) $ operator ","
+  args ← PC.sepBy (operator ":" *> (ident <|> anyKeyword)) $ operator ","
```

Comment:
Ah! Like "`foo`" is not ident, but `:in` is `operator ":" *> anyKeyword`! Nice catch